### ISSUE TITLE
Unify plotting syntax/logic

### ISSUE THREAD

**Issue by krischer** [obspy/obspy#1220]
As mentioned in #1219 we have lots of places in the code where a plot is/might be produced and the resulting figure can either be shown directly or be plotted to a passed figure object or saved to a file name. We should unify the logic and behavior of this across the code base.

Currently it is a bit like this:

* `plot` argument: Only given if a method/function does not produce a plot by default. If not given, all the following parameters will be ignored.
* `show` argument: Defaults to `True`. Shows the plot if given.
* `fig` argument: Plot to the given figure if passed. Otherwise create a new figure.
* `outfile` argument: Save to the file if given.
* `format` argument: File format. Only useful if `outfile` is a buffer of open file.

This is fairly ugly and the parameters interfere with each other.

Without breaking a lot of existing code we could do the following:

* For functions/methods that optionally produce plots: `plot=True/existing fig instance/filename,buffer`
* For functions that only produce a plot (this should require very few changes):
  * Always return the figure object.
  * Have the `show`, `fig`, `format`, and `outfile` arguments as always mainly to not break existing codes. 

Does anyone have a better solution for this?

**Comment by megies** [comment]
See also related #1156.

**Comment by MMesch** [comment]
I think the plot=True/fig/filename/buffer is missing one option. plot=True means plt.show() is always called. If With the single plot argument, there would be no option to plot and return fig = trace.remove_response(plot=True) without passing an existing fig. 

Maybe it is therefore useful to have a separate show keyword? Each plot function should then have plot and show arguments. The only arguments that are not necessary are outfile and fig, as well as format.

**Comment by megies** [comment]
> Maybe it is therefore useful to have a separate show keyword? Each plot function should then have plot and show arguments. The only arguments that are not necessary are outfile and fig, as well as format.

Sounds good to me.

**Comment by trichter** [comment]
3 developers, 5 opinions:

* plot option:
   * ``None``: default for processing functions, not available for plot functions -> do not plot
   * ``'show'``: default for plotting functions -> show
   * ``'return'``: do not show and return fig
   * ``fname``: do not show, write plot and and return fig
* ``fig`` option to pass a figure with non-default size and dpi
* ``format`` option? Is it necessary?
* ``close`` option? Is it necessary?

**Comment by krischer** [comment]
For 2.0.0 we can totally break things for this (at least a bit). Returning figures (while also showing them) has the nasty side effect of plotting them twice in jupyter notebooks if the plot call is the last call in a cell.

**Comment by trichter** [comment]
Then maybe it is best if a figure is not returned in this case?

**Comment by trichter** [comment]
I played a bit with decorators.

Here is the result

```py
import inspect
from functools import wraps
import matplotlib.pyplot as plt

FIGDOC = """:param fig: Figure"""
PLOTDOC1 = """:param plot: show, return, filename"""
PLOTDOC2 = """:param plot: None, show, return, filename"""


def plot(func):
    sig = inspect.signature(func)
    assert 'plot' in sig.parameters
    assert '{plotdoc}' in func.__doc__
    is_plot_func = sig.parameters['plot'].default is not None
    @wraps(func)
    def func2(*args, **kwargs):
        result = func(*args, **kwargs)
        plot = kwargs.get('plot', sig.parameters['plot'].default)
        if is_plot_func:
            if plot == 'show':
                fig = None
                plt.show()
            elif plot == 'return':
                fig = plt.gcf()
            else:
                fig = plt.gcf()
                fig.savefig(plot)
            return fig
        else:
            if plot is None:
                pass
            elif plot == 'show':
                plt.show()
            elif plot == 'return':
                result = result, plt.gcf()
            else:
                plt.savefig(plot)
            return result
    plotdoc = PLOTDOC1 if is_plot_func else PLOTDOC2
    func2.__doc__ = func2.__doc__.format(plotdoc=plotdoc, figdoc=FIGDOC)
    return func2


@plot
def plot_stuff(plot='show', fig=None):
    """
    plot stuff doc
    {plotdoc}
    {figdoc}
    """
    if fig is None:
        fig = plt.figure()
    plt.plot((1, 2), (1, 2))


@plot
def do_stuff(plot=None, fig=None):
    """
    do stuff doc
    {plotdoc}
    {figdoc}
    """
    if fig is None:
        fig = plt.figure()
    plt.plot((1, 2), (1, 2))
    return 1+1


print(plot_stuff.__doc__)
print()
print(do_stuff.__doc__)
print(plot_stuff(plot='show'))
print(do_stuff())
print(do_stuff(plot='return'))
```

Output:

```

    plot stuff doc
    :param plot: show, return, filename
    :param fig: Figure
    


    do stuff doc
    :param plot: None, show, return, filename
    :param fig: Figure
    
None
2
(2, <Figure size 640x480 with 1 Axes>)
```

**Comment by krischer** [comment]
I agree that going the decorator route is the simplest way to achieve consistency. I wonder if we should inject the function arguments, e.g. `fig`, ... as well but this is fairly hard to do fully dynamically. There is this beautiful package which aids in it but we don't really want another dependency: https://python-forge.readthedocs.io/en/latest/

**Comment by trichter** [comment]
Yep, it's doable and pretty cool, but on the other side its also quiet complex and probably not necessary? Also I think you use some flexibility without making it even more complex, e.g. for non-default defaults and default figures of different size. A check in the decorator might be enough to remind one of the necessary arguments.
